parlai/agents/fairseq/fairseq.py

```diff
@@ -107,7 +107,38 @@ def _fairseq_opt_wrapper(opt, skip_pretrained_embedding_loading=False):
 
 
 class _FairseqDictionary(DictionaryAgent):
-    """Skeleton dictionary class needed for interaction with fairseq-py"""
+    """
+    Skeleton dictionary class needed for interaction with fairseq-py.
+
+    This class mostly just adds some basic API behavior that Fairseq internally
+    expects from dictionaries.
+
+    It also inserts a fake token at the 0th index of the dictionary, as
+    fairseq-py maintains backwards compatibility with fairseq-lua, which uses
+    1 indexing.
+    """
+    # Name of our fake lua compatibility token
+    _LUA = '__LUACOMPAT__'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # insert the fairseq-lua compatibility token to emulate 1-indexing.
+        # This 1-indexing assumption is baked into a couple of places in fairseq-py,
+        # and is unavoidable at the moment.
+        #
+        # Because of the structure of DictionaryAgent, it's difficult to force
+        # a token in the 0th position without breaking load()ing. I've found
+        # this to be the best way.
+
+        # add the token to the dictionary
+        self.add_token(self._LUA)
+        self.freq[self._LUA] = self.freq[self.null_token] + 1
+        # Sort the list. Note that sort() calls remove_tail, so we need to
+        # ensure we don't throw anything away
+        tmp = self.minfreq
+        self.minfreq = -1
+        self.sort()
+        self.minfreq = tmp
```

Comment:
maybe better: let's add a kwarg to sort which says `trim=True`, which decides whether to call self.remove_tail as well as self.resize_to_max (which trims to the max number of total tokens). Then we can just do `self.sort(trim=False)` here.